extensions.bzl

```diff
@@ -14,35 +45,44 @@ def _doxygen_repository(ctx):
-        fail("The number of versions, sha256s and platforms must be the same")
+    if len(ctx.attr.versions) != len(ctx.attr.sha256s) or len(ctx.attr.versions) != len(ctx.attr.platforms) or len(ctx.attr.versions) != len(ctx.attr.executables):
+        fail("The number of versions, sha256s, platforms and executables must be the same")
     for platform in ctx.attr.platforms:
-        if platform not in ("windows", "mac", "linux"):
-            fail("Unsupported platform: '%s'. Available options are (windows, mac, linux)" % platform)
+        if platform not in ("windows", "mac", "mac-arm", "linux", "linux-arm"):
+            fail("Unsupported platform: '%s'. Available options are (windows, mac, mac-arm, linux, linux-arm)" % platform)
 
     ctx.file("WORKSPACE", "workspace(name = %s)\n" % repr(ctx.name))
     ctx.file("doxygen.bzl", ctx.read(ctx.attr.doxygen_bzl))
     ctx.file("BUILD.bazel", ctx.read(ctx.attr.build))
     ctx.file("Doxyfile.template", ctx.read(ctx.attr.doxyfile_template))
 
-    for doxygen_version, sha256, platform in zip(ctx.attr.versions, ctx.attr.sha256s, ctx.attr.platforms):
+    for doxygen_version, sha256, platform, executable in zip(ctx.attr.versions, ctx.attr.sha256s, ctx.attr.platforms, ctx.attr.executables):
+        if platform == "windows" and _get_current_platform(ctx) == platform:
+            executable_destination = "windows/doxygen.exe"
+        elif platform in ("mac", "mac-arm") and _get_current_platform(ctx) == platform:
+            executable_destination = "mac/doxygen"
+        elif platform in ("linux", "linux-arm") and _get_current_platform(ctx) == platform:
+            executable_destination = "linux/doxygen"
+        else:
+            continue  # Skip the platform if it does not match the current platform. No operation will be performed
+
         # If the version is set to 0.0.0 use the installed version of doxygen from the PATH
         # No download will be performed
         # This happens only if the platform matches the current platform
         if doxygen_version == "0.0.0":
-            if platform == "windows" and ctx.os.name.startswith("windows"):
-                ctx.file("windows/doxygen.exe", ctx.read(ctx.which("doxygen")), legacy_utf8 = False)
-            elif platform == "mac" and ctx.os.name.startswith("mac"):
-                ctx.file("mac/doxygen", ctx.read(ctx.which("doxygen")), legacy_utf8 = False)
-            elif platform == "linux" and ctx.os.name == "linux":
-                ctx.file("linux/doxygen", ctx.read(ctx.which("doxygen")), legacy_utf8 = False)
+            ctx.file(executable_destination, ctx.read(ctx.which("doxygen")), legacy_utf8 = False)
+            continue
+
+        # If an executable is provided, use it instead of downloading the doxygen binary
+        if executable != "":
+            ctx.file(executable_destination, ctx.read(executable), legacy_utf8 = False)
```

Comment:
User-provided executable now working on Linux!!

However, I'm currently wrestling with this error on macOS ARM:

```
ERROR: An error occurred during the fetch of repository 'rules_doxygen++doxygen_extension+doxygen':
   Traceback (most recent call last):
    File "/private/var/tmp/_bazel_kayce/64b8da36e1971548a86a520dd3605d75/external/rules_doxygen+/extensions.bzl", line 78, column 54, in _doxygen_repository
        ctx.file(executable_destination, ctx.read(executable), legacy_utf8 = False)
Error in read: java.io.FileNotFoundException: /home/kayce/.cache/bazel/_bazel_kayce/9659373b1552c281136de1c8eeb3080d/external/+_repo_rules+doxybin/mac-arm (No such file or directory)
```

Here are the relevant bits from my `MODULE.bazel`:

```
http_archive(
    name = "doxybin",
    build_file_content = """
package(default_visibility = ["//visibility:public"])
exports_files([
    "mac-arm",
    "mac-intel",
    "ubuntu-gcc",
    "windows.exe",
])
""",
    sha256 = "2624d7bb0b8c90ac604590f334f37c08ec05b5df7fb5fddc7e24d2c024f7d8af",
    strip_prefix = "doxybin-20241126.2",
    url = "https://github.com/kaycebasques/doxybin/archive/refs/tags/20241126.2.zip",
)

git_override(
    module_name = "rules_doxygen",
    commit = "4fe30cc55d1eba4d98a99e0478149b8f2c0ba590",
    remote = "https://github.com/tendto/rules_doxygen.git",
)

doxygen_extension = use_extension("@rules_doxygen//:extensions.bzl", "doxygen_extension")
doxygen_extension.configuration(
    executable = "@doxybin//:ubuntu-gcc",
    platform = "linux",
)
doxygen_extension.configuration(
    executable = "@doxybin//:mac-intel",
    platform = "mac",
)
doxygen_extension.configuration(
    executable = "@doxybin//:mac-arm",
    platform = "mac-arm",
)
doxygen_extension.configuration(
    executable = "@doxybin//:windows.exe",
    platform = "windows",
)
use_repo(doxygen_extension, "doxygen")
```

The fact that it's working on Linux but not macOS ARM is puzzling me. It seems like the executable does exist at the expected location.  E.g. when I manually download https://github.com/kaycebasques/doxybin/archive/refs/tags/20241126.2.zip and decompress that ZIP on my Mac the file is at the expected location. Gonna read through this file now and try to figure out if something is getting lost in translation here.

Reply:
OK, I think I figured it out. I think I'm back to a `MODULE.bazel.lock` problem. The `executables` are getting saved like this:

```
        "generatedRepoSpecs": {
          "doxygen": {
            "bzlFile": "@@rules_doxygen+//:extensions.bzl",
            "ruleClassName": "doxygen_repository",
            "attributes": {
              "versions": [
                "",
                "",
                "",
                "",
                "1.12.0"
              ],
              "sha256s": [
                "0000000000000000000000000000000000000000000000000000000000000000",
                "0000000000000000000000000000000000000000000000000000000000000000",
                "0000000000000000000000000000000000000000000000000000000000000000",
                "0000000000000000000000000000000000000000000000000000000000000000",
                "3c42c3f3fb206732b503862d9c9c11978920a8214f223a3950bbf2520be5f647"
              ],
              "platforms": [
                "linux",
                "mac",
                "mac-arm",
                "windows",
                "linux-arm"
              ],
              "executables": [
                "/home/kayce/.cache/bazel/_bazel_kayce/9659373b1552c281136de1c8eeb3080d/external/+_repo_rules+doxybin/ubuntu-gcc",
                "/home/kayce/.cache/bazel/_bazel_kayce/9659373b1552c281136de1c8eeb3080d/external/+_repo_rules+doxybin/mac-intel",
                "/home/kayce/.cache/bazel/_bazel_kayce/9659373b1552c281136de1c8eeb3080d/external/+_repo_rules+doxybin/mac-arm",
                "/home/kayce/.cache/bazel/_bazel_kayce/9659373b1552c281136de1c8eeb3080d/external/+_repo_rules+doxybin/windows.exe",
                ""
              ]
            }
          }
        },
```

The fact that the paths include my username was my clue that something was wrong. That path obviously won't exist for everyone else building Pigweed.

And then I realized why the `mac-arm` binary wasn't being found:

```
Error in read: java.io.FileNotFoundException: /home/kayce/.cache/bazel/_bazel_kayce/9659373b1552c281136de1c8eeb3080d/external/+_repo_rules+doxybin/mac-arm (No such file or directory)
```

`/home/kayce/...` is the path on my Linux machine whereas  when I run on macOS the path starts with `/private/var/...`

So I think it's close, I just need to figure out how to not store `executables` as literal paths in `MODULE.bazel.lock`